### ISSUE TITLE
watch scope

### ISSUE THREAD

**Issue by mvujica** [HerringtonDarkholme/av-ts#5]
Probably not an issue but a caveat or I'm doing something wrong.
Inside watch function type of (this) is Vue so I have to do something like this or it won't compile:

 ```
// Watched class property @Watch plus declaration and initiation
    @Watch(function (newValue, oldValue) {
        let self = <AddressComponent>this;
        if (newValue == 0)
            self.isNewAddress = true;
        else
            self.isNewAddress = false;
    })
    selectedBillingAddress: string = undefined;
```

also data and watched members must be initialized.

**Comment by HerringtonDarkholme** [comment]
Hi, this is limitation of TypeScript's inference schema. Watch's annotation cannot have the annotated class's information.

Try this:

```typescript
@Watch<Address, string>((newVal, oldVal) => {
})
```

**Comment by HerringtonDarkholme** [comment]
I might change the `Watch` function when https://github.com/Microsoft/TypeScript/pull/11929 is merged.

**Comment by HerringtonDarkholme** [comment]
New style watch in next branch
https://github.com/HerringtonDarkholme/av-ts/tree/next

**Comment by mvujica** [comment]
Ok, next branch has it stringly typed like other libraries and vue, but I really like your strongly typed style, although it require additional annotation for "this" scope.

**Comment by HerringtonDarkholme** [comment]
Hi, it has a string API but everything is type checked! It's powered by typescript's new `keyof` and mapped type. We can achieve type safety without additional annotation

**Comment by mvujica** [comment]
Really nice, when do you plan to merge?

**Comment by HerringtonDarkholme** [comment]
Probably after typescript officially released mapped types in stable.

**Comment by whitetrefoil** [comment]
Watching on this issue...
Using `this.$watch` in `mounted()` until new TS release...

**Comment by HerringtonDarkholme** [comment]
Shipped in 0.6.0!
